packages/daemon/src/pet-store.js

```diff
@@ -10,228 +12,238 @@ const validFormulaPattern =
+
+    const follow = async () =>
+      makeIteratorRef(
+        (async function* currentAndSubsequentNames() {
+          const changes = changesTopic.subscribe();
+          for (const name of [...petNames.keys()].sort()) {
+            yield { add: name };
+          }
+          yield* changes;
+        })(),
+      );
 
-    const petNamePath = powers.joinPath(petNameDirectoryPath, petName);
-    const petNameText = `${formulaIdentifier}\n`;
-    await powers.writeFileText(petNamePath, petNameText);
-    changesTopic.publisher.next({ add: petName });
-  };
+    /**
+     * @param {string} petName
+     */
+    const remove = async petName => {
+      assertPetName(petName);
+      const formulaIdentifier = petNames.get(petName);
+      if (formulaIdentifier === undefined) {
+        throw new Error(
+          `Formula does not exist for pet name ${JSON.stringify(petName)}`,
+        );
+      }
+      if (!validFormulaPattern.test(formulaIdentifier)) {
+        throw new Error(`Invalid formula identifier ${q(formulaIdentifier)}`);
+      }
 
-  const list = () => harden([...petNames.keys()].sort());
+      const petNamePath = filePowers.joinPath(petNameDirectoryPath, petName);
+      await filePowers.removePath(petNamePath);
+      petNames.delete(petName);
+      const formulaPetNames = formulaIdentifiers.get(petName);
+      if (formulaPetNames !== undefined) {
+        formulaPetNames.delete(petName);
+      }
+      changesTopic.publisher.next({ remove: petName });
+      // TODO consider retaining a backlog of deleted names for recovery
+      // TODO consider tracking historical pet names for formulas
+    };
+
+    /**
+     * @param {string} fromName
+     * @param {string} toName
+     */
+    const rename = async (fromName, toName) => {
+      assertPetName(fromName);
+      assertPetName(toName);
+      if (fromName === toName) {
+        return;
+      }
+      const formulaIdentifier = petNames.get(fromName);
+      const overwrittenFormulaIdentifier = petNames.get(toName);
+      if (formulaIdentifier === undefined) {
+        throw new Error(
+          `Formula does not exist for pet name ${JSON.stringify(fromName)}`,
+        );
+      }
+      if (!validFormulaPattern.test(formulaIdentifier)) {
+        throw new Error(`Invalid formula identifier ${q(formulaIdentifier)}`);
+      }
+      if (
+        overwrittenFormulaIdentifier !== undefined &&
+        !validFormulaPattern.test(overwrittenFormulaIdentifier)
+      ) {
+        throw new Error(
+          `Invalid formula identifier ${q(overwrittenFormulaIdentifier)}`,
+        );
+      }
 
-  const follow = async () =>
-    makeIteratorRef(
-      (async function* currentAndSubsequentNames() {
-        const changes = changesTopic.subscribe();
-        for (const name of [...petNames.keys()].sort()) {
-          yield { add: name };
+      const fromPath = filePowers.joinPath(petNameDirectoryPath, fromName);
```

Comment:
I look forward to abstracting file powers even more thoroughly. This might mean that the daemon powers consist of a content store, and a “tables” for pet names, mailboxes, and formulas.